ca-rotation-script/CA-Rotation-CAPI.md

```diff
@@ -0,0 +1,108 @@
+# Introduction
+The purpose of this document is to provide step-by-step instructions how to do CA rotation on CAPI. For the remainder of this document, we use Kind as the management cluster and Cluster API Provider Docker (CAPD) as the infrastructure provider.
+- Manually rotate CA in the workload cluster: [1][1] 
+- Update the management (CAPI) cluster secrets so it can interact with the workload cluster without any TLS problems.
+
+# Prerequisites
+- Go version 1.13.0+
+- Docker
+- Kind v0.6 or v0.7 (other clusters can be used if preload_images_for_kind is set to false)
+
+# How to do
+## Setup the management cluster and workload cluster
+Firstly, install `clusterctl`. `v0.3.14` is used in the experiments in this document. The instruction can be found in https://cluster-api.sigs.k8s.io/user/quick-start.html
+```sh
+clusterctl version # Ensure that clusterctl is installed correctly
+```
+Then, create a kind cluster:
+```sh
+cat > kind-cluster-with-extramounts.yaml <<EOF
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- role: control-plane
+  extraMounts: # This mount ensures that any CAPI pods inside this kind cluster can interact with the docker socker of the host
+    - hostPath: /var/run/docker.sock
+      containerPath: /var/run/docker.sock
+EOF
+
+kind create cluster --config kind-cluster-with-extramounts.yaml
+```
+Next, initialize the management cluster by transforming the kind cluster into a management cluster:
+```sh
+clusterctl init --infrastructure docker
+```
+After that, create the workload cluster:
+```sh
+kubectl apply -f self-hosted-cluster-template.yaml
+```
+Note that if some objects cannot be created, that means the management cluster is not fully initialized. Wait for a few seconds then try the above command again. 
+
+Next, extract the kubeconfig of the workload cluster:
+```sh
+clusterctl get kubeconfig my-cluster > capi-kubeconfig
+```
+In this example, we deploy Calico CNI solution. Deploy Calico using capi-kubeconfig copied in the last step with:
+```sh
+kubectl --kubeconfig=capi-kubeconfig apply -f https://docs.projectcalico.org/v3.15/manifests/calico.yaml
+```
+Then, wait until all KCP and worker nodes are ready:
+```sh
+$ kubectl get kcp
+NAME              READY   INITIALIZED   REPLICAS   READY REPLICAS   UPDATED REPLICAS   UNAVAILABLE REPLICAS
+my-controlplane           true          1          1                1                  
+
+$ kubectl get machinedeployment
+NAME     PHASE     REPLICAS   AVAILABLE   READY
+worker   Running   1          1           1
+```
+## Pivoting
+If we want to make sure that the CA rotation procedure works with the case that the cluster is self-hosted, we have to do pivoting.
+
+Firstly, transform the workload cluster into another management cluster:
+```sh
+clusterctl init --kubeconfig capi-kubeconfig --infrastructure docker
+```
+Next, move all CAPI components from the original management cluster to the new management cluster:
```

Comment:
Is it moving only CAPI components or rather infrastructure-provider components also?